### ISSUE TITLE
image promoter authorization failure

### ISSUE THREAD

**Issue by listx** [kubernetes/k8s.io#426]
Failed promotion due to authorization failure: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/post-k8sio-cip/1187747140168847360

Snipped logs below:
```
Activated service account credentials for: [k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com]
Built:   2019-10-21 16:02:36+0000
Version: v2.3.1-34-ge32056e
Commit:  e32056e32fc25e0e371ba78066cb79b3a21bebd5
...
<snipped>
...
I1025 15:06:32.664314      12 inventory.go:1884] Pending promotions:
I1025 15:06:32.664373      12 inventory.go:1886]   {{gcr.io/k8s-staging-capi-kubeadm   true} {cluster-api-kubeadm-controller v0.1.5} sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf {us.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com  false} {cluster-api-kubeadm-controller v0.1.5}}
I1025 15:06:32.664409      12 inventory.go:1886]   {{gcr.io/k8s-staging-capi-kubeadm   true} {cluster-api-kubeadm-controller v0.1.5} sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf {asia.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com  false} {cluster-api-kubeadm-controller v0.1.5}}
I1025 15:06:32.664500      12 inventory.go:1886]   {{gcr.io/k8s-staging-capi-kubeadm   true} {cluster-api-kubeadm-controller v0.1.5} sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf {eu.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com  false} {cluster-api-kubeadm-controller v0.1.5}}
I1025 15:06:32.664634      12 inventory.go:1700] ---------- BEGIN PROMOTION ----------
E1025 15:06:32.885779      12 inventory.go:1935] failed to copy index: UNAUTHORIZED: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication
E1025 15:06:32.885824      12 inventory.go:1467] Request {{0 gcr.io/k8s-staging-capi-kubeadm us.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com cluster-api-kubeadm-controller cluster-api-kubeadm-controller sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf  v0.1.5} <nil>}: error(s) encountered: [{running writeImage() failed to copy index: UNAUTHORIZED: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication}]
E1025 15:06:32.897892      12 inventory.go:1935] failed to copy index: UNAUTHORIZED: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication
E1025 15:06:32.897918      12 inventory.go:1467] Request {{0 gcr.io/k8s-staging-capi-kubeadm eu.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com cluster-api-kubeadm-controller cluster-api-kubeadm-controller sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf  v0.1.5} <nil>}: error(s) encountered: [{running writeImage() failed to copy index: UNAUTHORIZED: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication}]
E1025 15:06:32.928442      12 inventory.go:1935] failed to copy index: UNAUTHORIZED: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication
```

Promoter version: v2.3.1-34-ge32056e
Service account used: k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com

**Comment by listx** [comment]
The go-containerregistry library used in the version above succeeded in e2e tests here: https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/131 (direct link: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_k8s-container-image-promoter/131/pull-cip-e2e/1174895151210827776). The promoter version used in that test was Version: v2.3.0-4-g878fb9e, which still incorporates the change to use go-containerregistry for image copies that were introduced in https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/118/commits/522e885080643b991cde38bfe65abfa6efc4b5e5. This leads me to think that this is a credential issue (somehow, somewhere).

/cc @thockin

**Comment by listx** [comment]
From the maintainer of go-containerregistry, on slack: (@jonjohnson)
```
gcrane is just using https://godoc.org/golang.org/x/oauth2/google#FindDefaultCredentials to get the ADC, I'd be really surprised if there's a bug there, but maybe I'm using it wrong
```

**Comment by listx** [comment]
Below is a paste of a small script i wrote as a sanity check. It uses a new json key for the same SA. It tested 3 things in order:

(1) FAIL: gcrane binary built locally from bb2eda47dcc0c4cb5d554415f786c8fbe45e42fb (the same sha that is referenced in the promoter's WORKSPACE file in bazel)
(2) FAIL: gcrane docker image built recently (earlier this week) from 7683b4ee5f6150cb47a791309f781c522b95a58f
(3) WORKS: same as (1), but authenticating first as gcloud and *not* passing in Google ADC env var.

Before I pollute the discussion with my weak understanding of creds, @jonjohnsonjr do you have any ideas?

```
#!/usr/bin/env bash

echo
echo ----
echo "using old gcrane version built from bb2eda47dcc0c4cb5d554415f786c8fbe45e42fb of go-containerregistry"
GOOGLE_APPLICATION_CREDENTIALS=~/secure/k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com.json \
    gcrane cp \
    "gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf" \
    "us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller:v0.1.5"

echo
echo ----
echo "using gcrane version built from 7683b4ee5f6150cb47a791309f781c522b95a58f of go-containerregistry"
docker run -v ~/secure/k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com.json:/auth.json \
    --env GOOGLE_APPLICATION_CREDENTIALS=/auth.json \
    gcr.io/go-containerregistry/gcrane@sha256:560757a7b63c85e9b95d2d971a18aa2a2425899e36f1550ce3584018d9ac49ea cp \
    "gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf" \
    "us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller:v0.1.5"

echo
echo ----
echo "using old gcrane version built from bb2eda47dcc0c4cb5d554415f786c8fbe45e42fb of go-containerregistry, but with gcloud auth instead of ADC"
gcloud auth activate-service-account --key-file ~/secure/k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com.json

gcrane cp \
    "gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf" \
    "us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller:v0.1.5"

# RESULTS

----
using old gcrane version built from bb2eda47dcc0c4cb5d554415f786c8fbe45e42fb of go-containerregistry
2019/10/25 11:44:06 Pulling gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf
2019/10/25 11:44:07 Pushing us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller:v0.1.5
2019/10/25 11:44:07 failed to copy image: failed to copy index: DENIED: Token exchange failed for project 'k8s-artifacts-prod'. Caller does not have permission 'storage.buckets.get'. To configure permissions, follow instructions at: https://cloud.google.com/container-registry/docs/access-control

----
using gcrane version built from 7683b4ee5f6150cb47a791309f781c522b95a58f of go-containerregistry
2019/10/25 18:44:08 Pulling gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf
2019/10/25 18:44:08 No matching credentials were found, falling back on anonymous
2019/10/25 18:44:08 Pushing us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller:v0.1.5
2019/10/25 18:44:08 No matching credentials were found, falling back on anonymous
2019/10/25 18:44:09 failed to copy image: failed to copy index: unsupported status code 404; body: <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 404 (Not Found)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL <code>/v2/token</code> was not found on this server.  <ins>That’s all we know.</ins>

----
using old gcrane version built from bb2eda47dcc0c4cb5d554415f786c8fbe45e42fb of go-containerregistry, but with gcloud auth instead of ADC
Activated service account credentials for: [k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com]


Updates are available for some Cloud SDK components.  To install them,
please run:
  $ gcloud components update



To take a quick anonymous survey, run:
  $ gcloud alpha survey

2019/10/25 11:44:10 Pulling gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf
2019/10/25 11:44:10 Pushing us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller:v0.1.5
2019/10/25 11:44:12 existing blob: sha256:e8d8785a314f385d3675a017f4e2df1707c528c06e7a7989663fdab4900bd8ff
2019/10/25 11:44:14 pushed blob: sha256:cdb4c601461393cfae1bee3cd1bbca9e83dcce13b259602268d5680791bb18c9
2019/10/25 11:44:16 pushed blob: sha256:0239516990bed21b2c4347b268f1b30e42fd81af9a328ba990546d65d0ade35b
2019/10/25 11:44:16 us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller@sha256:37b1c1c1dc4b1eaa4f6b8992542172b1e42fb67d2f2f7db1732a0f14fd7d8a0c: digest: sha256:37b1c1c1dc4b1eaa4f6b8992542172b1e42fb67d2f2f7db1732a0f14fd7d8a0c size: 739
2019/10/25 11:44:17 existing blob: sha256:e8d8785a314f385d3675a017f4e2df1707c528c06e7a7989663fdab4900bd8ff
2019/10/25 11:44:20 pushed blob: sha256:c9176fb48226ac96a7659bee6157cb28835709de3df3513e623883554666557e
2019/10/25 11:44:21 pushed blob: sha256:10d4ce4848b73f9ea209f72e963327018d0edc0713c9c8e5cfa3216a9358a489
2019/10/25 11:44:22 us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller@sha256:a3bd7c8030e956b3464b069281dc854be53a436430c731940e856e2802e25231: digest: sha256:a3bd7c8030e956b3464b069281dc854be53a436430c731940e856e2802e25231 size: 739
2019/10/25 11:44:22 existing blob: sha256:e8d8785a314f385d3675a017f4e2df1707c528c06e7a7989663fdab4900bd8ff
2019/10/25 11:44:25 pushed blob: sha256:ad3f172e8b57b0a3c8e8c2f3a4e6d439376868539a17bedb4d579d6d36316c25
2019/10/25 11:44:26 pushed blob: sha256:9e72c047786f4709d84d181c9c8d9ebf52e1a2b8a71ad012987ab33104f89be3
2019/10/25 11:44:26 us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller@sha256:5f1b66a4e327952d06b56cc9b86569368573bfa421d85b5a0cff7855da45164d: digest: sha256:5f1b66a4e327952d06b56cc9b86569368573bfa421d85b5a0cff7855da45164d size: 739
2019/10/25 11:44:27 existing blob: sha256:e8d8785a314f385d3675a017f4e2df1707c528c06e7a7989663fdab4900bd8ff
2019/10/25 11:44:30 pushed blob: sha256:594d6a1a43e2015e9104c03bb5b4caa4516c0e92f3d9ee64c19d99543acbaba5
2019/10/25 11:44:30 pushed blob: sha256:17e2850ce7b1ea30ecd8672574ec67be600e000af208eef33551df452d249890
2019/10/25 11:44:31 us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller@sha256:2dffbfb8ce6cf49ceca8a5cd836b1fe21646ab71bf378113337d4b549c719e8d: digest: sha256:2dffbfb8ce6cf49ceca8a5cd836b1fe21646ab71bf378113337d4b549c719e8d size: 739
2019/10/25 11:44:32 existing blob: sha256:e8d8785a314f385d3675a017f4e2df1707c528c06e7a7989663fdab4900bd8ff
2019/10/25 11:44:34 pushed blob: sha256:0629b7f8e3568cb08910c298d1c4bc3a2f43ff0514ed07178e4ce22636d1c380
2019/10/25 11:44:35 pushed blob: sha256:2c725f9e919a2d11c679c7f489c38ff336ecde0bb4a64eeb72ec5e7d73b86877
2019/10/25 11:44:36 us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller@sha256:ccd0ac81873cd375e0b70e7b14580a0c2ad79394bf11e64cd5344898470a0355: digest: sha256:ccd0ac81873cd375e0b70e7b14580a0c2ad79394bf11e64cd5344898470a0355 size: 739
2019/10/25 11:44:36 us.gcr.io/k8s-artifacts-prod/capi-kubeadm/cluster-api-kubeadm-controller:v0.1.5: digest: sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf size: 1665
```

**Comment by listx** [comment]
I guess the troubling thing is that the gcrane binary i built from the same version that is used by the promoter fails to push blobs if i invoke google adc, but works if i do not invoke adc (with same SA all the way across)

**Comment by listx** [comment]
This might be related to https://github.com/google/go-containerregistry/issues/593 which should be fixed by https://github.com/google/go-containerregistry/pull/594

**Comment by listx** [comment]
I am going to create a new version of the promoter locally incorporating this fix to attempt to promote the current state of the promoter manifests (invoke as a docker image of course). I can do this later today.

**Comment by listx** [comment]
I just did a test by pulling in go-containerregistry version with the fix in google/go-containerregistry#594 here: https://github.com/listx/k8s-container-image-promoter/tree/auth-fix-attempt, building the image locally and running this script:

```
#!/usr/bin/env bash

img="gcr.io/k8s-staging-artifact-promoter/cip:20191025-v2.3.1-35-g55a6c18"
echo "-----------------------"
docker run --rm -it $img "cip -version"
echo "-----------------------"

read -r -d '' CMD <<EOF
cd /go/src/github.com/kubernetes/k8s.io &&
cip \
 -thin-manifest-dir=/go/src/github.com/kubernetes/k8s.io/k8s.gcr.io \
 -key-files=/secure/k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com.json \
 -dry-run=false
EOF

docker run \
    --rm \
    -it \
    -v $HOME/secure:/secure \
    -v $HOME/go:/go \
    -e "GOOGLE_APPLICATION_CREDENTIALS=/secure/k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com.json" \
    $img \
    "$CMD"
```

and it gave me this error (different error this time, although maybe that's just the library's error msg formatting that changed):

```
<snipped>
I1025 20:02:45.114775       1 cip.go:341] ********** FINISHED **********
E1025 20:02:45.114777       1 inventory.go:1467] Request {{0 gcr.io/k8s-staging-capi-kubeadm eu.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com cluster-api-kubeadm-controller cluster-api-kubeadm-controller sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf  v0.1.5} <nil>}: error(s) encountered: [{running writeImage() failed to copy index: unsupported status code 404; body: <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 404 (Not Found)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL <code>/v2/token</code> was not found on this server.  <ins>That’s all we know.</ins>
}]
```

Doing it without the env var to Google ADC appears to result in the same error.

**Comment by jonjohnsonjr** [comment]
> I guess the troubling thing is that the gcrane binary i built from the same version that is used by the promoter fails to push blobs if i invoke google adc, but works if i do not invoke adc (with same SA all the way across)

Sounds like it would be falling back to gcloud?

> I just did a test by pulling in go-containerregistry version with the fix in google/go-containerregistry#594 here: https://github.com/listx/k8s-container-image-promoter/tree/auth-fix-attempt, building the image locally and running this script:

That fix is specifically for gcrane. How credentials are plumbed through the promoter is dependent on a lot of things. How does the promoter image fetch credentials? What's the `~/.docker/config.json` look like and how is it populated? I'm not sure how gcloud credentials work, but if

**Comment by listx** [comment]
OK so I found a workaround. It involves simply populating the `~/.docker/config.json`file with

```
{
  "credHelpers": {
    "gcr.io": "gcloud", 
    "us.gcr.io": "gcloud", 
    "eu.gcr.io": "gcloud", 
    "asia.gcr.io": "gcloud", 
    "staging-k8s.gcr.io": "gcloud", 
    "marketplace.gcr.io": "gcloud"
  }, 
  "experimental": "enabled"
}
```

(I just copy-pasted from what I have in my local setup). An interesting side effect I observed is that if I have a blank file called `~/.docker/config.json` that is an empty file, I'll get an error like:

```
I1025 21:17:33.874024       1 inventory.go:1700] ---------- BEGIN PROMOTION ----------
E1025 21:17:33.874519       1 inventory.go:1935] fetching "gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf": /.docker/config.json: EOF
E1025 21:17:33.874577       1 inventory.go:1467] Request {{0 gcr.io/k8s-staging-capi-kubeadm asia.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com cluster-api-kubeadm-controller cluster-api-kubeadm-controller sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf  v0.1.5} <nil>}: error(s) encountered: [{running writeImage() fetching "gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf": /.docker/config.json: EOF}]
E1025 21:17:33.874769       1 inventory.go:1935] fetching "gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf": /.docker/config.json: EOF
E1025 21:17:33.874863       1 inventory.go:1467] Request {{0 gcr.io/k8s-staging-capi-kubeadm eu.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com cluster-api-kubeadm-controller cluster-api-kubeadm-controller sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf  v0.1.5} <nil>}: error(s) encountered: [{running writeImage() fetching "gcr.io/k8s-staging-capi-kubeadm/cluster-api-kubeadm-controller@sha256:669f1661513b8b250e56bc0b519f8873c83bd65ffb3087e0139a0b6b95a69cbf": /.docker/config.json: EOF}]
I1025 21:17:33.874908       1 cip.go:341] ********** FINISHED **********
```

which implies that having this file is important if I'm falling back to gcloud.

This still leaves open the question of "how did the e2e tests ever work?". I think it's because... the e2e tests we have [do this cred handling logic for us](https://github.com/kubernetes-sigs/k8s-container-image-promoter/blob/master/test-e2e/e2e-entrypoint-from-container.sh#L31-L37) as preparation for the test!!! (My bad, I should have thought of this...).

So the fix is: (1) add these cred helper defaults to the promoter image itself, and (2) remove these helpers for the e2e test entrypoint so that we don't have this whole scenario play itself out again in the future due to a mismatch of how creds and handled.
